### PR TITLE
Add methods to determine Vesein Tal Umatar/Vesein Berachah...

### PR DESCRIPTION
**Pull request by plonibarploni**
Add methods to determine Vesein Tal Umatar/Vesein Berachah, Mashiv Haruach/Hatal

ported from https://github.com/pinnymz/ruby-zmanim
relevant commits:
https://github.com/pinnymz/ruby-zmanim/commit/501e4f4bbef57ece0dfa17279b41ca3e0be96045
https://github.com/pinnymz/ruby-zmanim/commit/5932c10666a3e823a5f34c9c6145f4585ff74f7f

### CODE REVIEW

**Comment by KosherJava** [comment]
@plonibarploni , please update this to account for the package change to com.kosherjava.zmanim.
Thanks

**Comment by plonibarploni** [comment]
@KosherJava I rebased and then committed it again.

**Comment by yparitcher** [comment]
Btw, for a more complete implementation of Vesein Tal Umatar/Vesein Berachah that accounts for the Julian / Gregorian change, based on shulcan aruch siman 117 see: https://github.com/yparitcher/libzmanim/blob/65d0ba20d98e95a771fb708ad2e58eb91ff6aae6/src/hebrewcalendar.c#L883-L930

**Comment by KosherJava** [comment]
> Btw, for a more complete implementation of Vesein Tal Umatar/Vesein Berachah that accounts for the Julian / Gregorian change, based on shulchan aruch siman 117 see: https://github.com/yparitcher/libzmanim/blob/65d0ba20d98e95a771fb708ad2e58eb91ff6aae6/src/hebrewcalendar.c#L883-L930

Thank you @yparitcher . We certainly should enhance the patch with that. @plonibarploni , is this something that you can pick up?

**Comment by plonibarploni** [comment]
I've been very busy with work and it doesn't seem like I will have time to work on this for the next week or two, so you're welcome to work on it if you'd like.

**Comment by yparitcher** [comment]
This should work, mostly just C -> Java, I did not test as i am too lazy to install maven/gradle.

```patch
diff --git a/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java b/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java
index 1aa3c5a..8b67e3b 100644
--- a/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java
+++ b/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java
@@ -240,6 +240,54 @@ public class JewishCalendar extends com.kosherjava.zmanim.hebrewcalendar.JewishD
 		return false;
 	}
 
+	/**
+	 * Tekufas Shmuel: a solar year is 365.25 days.
+	 * notation: days,hours,chalakim
+	 * molad BaHaRad was 2D,5H,204C
+	 * or 5H,204C from the start of rosh hashana year 1
+	 * molad nissan add 177D,4H,438C (6 * 29D,12H,793C)
+	 * or 177D,9H,642C after rosh hashana year 1
+	 * tekufas nissan was 7D,9H,642C before molad nissan ~rambam.
+	 * or 170D,0H,0C after rosh hashana year 1
+	 * tekufas tishrei was 182D,3H (365.25 / 2) before tekufas nissan
+	 * or 12D,15H before Rosh Hashana year 1
+	 * outside of EY we say תל ומטר in ברכת השנים from 60 days after tekufas tishrei.
+	 * 60 includes the day of the tekufah and the day we start.
+	 * 60 days from the tekufah == 47D,9H from Rosh Hashana year 1
+	 */
+	int tekufasTishreiElapsedDays()
+	{
+		// days since Rosh Hashana year 1
+		// add 1/2 day as the first tekufas tishrei was 9 hours into the day
+		// this allows all 4 years of the secular leap year cycle to share 47 days
+		// make from 47D,9H to 47D for simplicity
+		double days = getJewishCalendarElapsedDays(getJewishYear()) + (getDaysSinceStartOfJewishYear()-1) + .5;
+		// days of completed solar years
+		double solar = (getJewishYear()-1)*365.25;
+		return (int) days - solar;
+	}
+
+	public boolean isBirchasHashanimToday()
+	{
+		if (getInIsrael() && getJewishMonth() == 7 && getJewishDayOfMonth() == 7) {return true;}
+		else if (tekufasTishreiElapsedDays() == 47) {return true;}
+		return false;
+	}
+
+	public boolean getBirchasHashanim()
+	{
+		if (getJewishMonth() == 1 && getJewishDayOfMonth() < 15) {return true;}
+		if (getJewishMonth() < 7) {return 0;}
+		if (getInIsrael())
+		{
+			if (getJewishMonth() == 7 && getJewishDayOfMonth() < 7) {return false;}
+			else {return true;}
+		} else {
+			if (tekufasTishreiElapsedDays() < 47) {return false;}
+			else {return true;}
+		}
+	}
+
 	/**
 	 * Return the type of year for parsha calculations. The algorithm follows the
 	 * <a href="http://hebrewbooks.org/pdfpager.aspx?req=14268&amp;st=&amp;pgnum=222">Luach Arba'ah Shearim</a> in the Tur Ohr Hachaim.
```

**Comment by plonibarploni** [comment]
@KosherJava Sorry it's been quite hectic here, but I just added @yparitcher's implementation.

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java [hunk 1]
```diff
@@ -240,6 +240,47 @@ public boolean isBirkasHachamah() {
...
+	 * or 177D,9H,642C after rosh hashana year 1
+	 * tekufas nissan was 7D,9H,642C before molad nissan ~rambam.
+	 * or 170D,0H,0C after rosh hashana year 1
+	 * tekufas tishrei was 182D,3H (365.25 / 2) before tekufas nissan
+	 * or 12D,15H before Rosh Hashana year 1
+	 * outside of EY we say תל ומטר in ברכת השנים from 60 days after tekufas tishrei.
+	 * 60 includes the day of the tekufah and the day we start.
+	 * 60 days from the tekufah == 47D,9H from Rosh Hashana year 1
+	 */
+	public int tekufasTishreiElapsedDays() {
```
Comment (KosherJava) [inline src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java:258]: Thank you @plonibarploni.  Should this be public?

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java [hunk 2]
```diff
@@ -240,6 +240,47 @@ public boolean isBirkasHachamah() {
...
+	 * notation: days,hours,chalakim
+	 * molad BaHaRad was 2D,5H,204C
+	 * or 5H,204C from the start of rosh hashana year 1
+	 * molad nissan add 177D,4H,438C (6 * 29D,12H,793C)
+	 * or 177D,9H,642C after rosh hashana year 1
+	 * tekufas nissan was 7D,9H,642C before molad nissan ~rambam.
+	 * or 170D,0H,0C after rosh hashana year 1
+	 * tekufas tishrei was 182D,3H (365.25 / 2) before tekufas nissan
+	 * or 12D,15H before Rosh Hashana year 1
+	 * outside of EY we say תל ומטר in ברכת השנים from 60 days after tekufas tishrei.
```
Comment (KosherJava) [inline src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java:254]: @plonibarploni Thanks. Please chnage the Hebrew chars to Birkas Hashanim and Tal Umatar. The charset for the files are not gauranteed to be UTF-8 everywhere.

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java (hunk 1 above)
Comment (plonibarploni) [reply src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java:258]: I can't think of any reason why it shouldn't be.

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java [hunk 3]
```diff
@@ -933,6 +974,58 @@ public Daf getDafYomiYerushalmi() {
...
 
+	public boolean mashivHaruachStarts() {
+		return getJewishMonth() == 7 && getJewishDayOfMonth() == 22;
+	}
+
+	public boolean mashivHaruachEnds() {
+		return getJewishMonth() == 1 && getJewishDayOfMonth() == 15;
+	}
+
+	public boolean mashivHaruach() {
```
Comment (KosherJava) [inline src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java:985]: @plonibarploni , please rename to isMashivHaruach()

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java (hunk 2 above)
Comment (plonibarploni) [reply src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java]: Fixed

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java (hunk 3 above)
Comment (plonibarploni) [reply src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java]: Done

**Comment by KosherJava** [comment]
@plonibarploni there are still a few minor items. can you address them?
Thanks

File: src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java (hunk 1 above)
Comment (plonibarploni) [reply src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java]: I changed it to private

**Comment by plonibarploni** [comment]
I believe I have fixed the last outstanding review.

**Review by KosherJava — APPROVED** [review]
Looks good. Thank you @plonibarploni  and @yparitcher

**Comment by yparitcher** [comment]
@KosherJava 
This PR adds 2 slightly conflicting calculations for the same thing (vsein tal umatar levracha / sheilag geshamim).

This is discussed in סימן קי"ז where (for Chutz laretz) we start 60 after the tekufah. The tekufah is figured according to Shmuel's opinion that a Solar year is 365.25 days (The same as the Julian Calendar).  For this reason the בית יוסף writes that it will always be november 22/23 as during his time they still used the Julian calendar. Due to the slight differences between the Gregorian & Julian calendar between the years 1900 - 2100 we can simplify and say it is always (the night after) December 4/5.

with this PR we have `isVesainTalUmatarLivrachaRecited()` which calculates based on the tekufah and `isVeseinTalUmatarRecited` which always uses Dec. 4/5. This is confusing for a user and needs to be clarified.
Also is there an opinion that holds it is always Dec. 4/5? If yes the 2 implementations should be put next to each other and named accordingly, and  if not we can use the more acurate method and drop the Dec. 4/5.

N.B.
Currently the 2 calculations return the same date and will do so until 2100.
The Julian & Gregorian calendar currently are different by 13 days, Nov 22/23 + 13 days = Dec 5/6 or the night after Dec 4/5.

**Comment by yparitcher** [comment]
The sefer  טעמי המנהגים has a arichus in this [HebrewBooks link](https://www.hebrewbooks.org/pdfpager.aspx?req=14556&st=&pgnum=372&hilite=) Siman תתכז in the הערה. (it continues on the next page)

**Comment by KosherJava** [comment]
> @KosherJava This PR adds 2 slightly conflicting calculations for the same thing (vsein tal umatar levracha / sheilag geshamim).
> 
> This is discussed in סימן קי"ז where (for Chutz laretz) we start 60 after the tekufah. The tekufah is figured according to Shmuel's opinion that a Solar year is 365.25 days (The same as the Julian Calendar). For this reason the בית יוסף writes that it will always be november 22/23 as during his time they still used the Julian calendar. Due to the slight differences between the Gregorian & Julian calendar between the years 1900 - 2100 we can simplify and say it is always (the night after) December 4/5.
> 
> with this PR we have `isVesainTalUmatarLivrachaRecited()` which calculates based on the tekufah and `isVeseinTalUmatarRecited` which always uses Dec. 4/5. This is confusing for a user and needs to be clarified. Also **is there an opinion that holds it is always Dec. 4/5?** If yes the 2 implementations should be put next to each other and named accordingly, and if not we can use the more acurate method and drop the Dec. 4/5.
> 
> N.B. Currently the 2 calculations return the same date and will do so until 2100. The Julian & Gregorian calendar currently are different by 13 days, Nov 22/23 + 13 days = Dec 5/6 or the night after Dec 4/5.

Thank you @yparitcher . The latest code should address your valid points. Please let me know if there are still any issues.

**Comment by yparitcher** [comment]
https://github.com/KosherJava/zmanim/blob/1f22ecce5573f856c6d2aba81627bd4e759d6c2e/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java#L335-L390

Should probably be moved down with the other season based zmanim.

https://github.com/KosherJava/zmanim/blob/1f22ecce5573f856c6d2aba81627bd4e759d6c2e/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java#L1164-L1169
and
https://github.com/KosherJava/zmanim/blob/1f22ecce5573f856c6d2aba81627bd4e759d6c2e/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java#L1181-L1191
are not needed.


Other than that it looks good

**Comment by yparitcher** [comment]
and

https://github.com/KosherJava/zmanim/blob/1f22ecce5573f856c6d2aba81627bd4e759d6c2e/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java#L366
should be
```java
		if (inIsrael && getJewishMonth() == CHESHVAN && getJewishDayOfMonth() == 7) {
```

**Comment by KosherJava** [comment]
@yparitcher , thanks. getJewishVeseinTalUmatarStartDate() was used in the code, and I converted it to private. I feel that there is value in isVeseinTalUmatarStartsTonight() so that someone generating a calendar can post a note that it starts on this night. What are your thoughts?

**Comment by yparitcher** [comment]
```
public boolean isVeseinTalUmatarStartsTonight() {
		if (getDayOfWeek() == 6) { return false;}
		if (inIsrael) {
			if (getJewishMonth() == CHESHVAN) {
				if (getJewishDayOfMonth() == 7 || (getDayOfWeek() == 7 && getJewishDayOfMonth() == 8)) {return true;}
			}
			return false;
		} else {
			return getTekufasTishreiElapsedDays() == 46 || (getDayOfWeek() == 7 && getTekufasTishreiElapsedDays() == 47);
		}
	}
```

**Comment by yparitcher** [comment]
I just realized that ` isVeseinTalUmatarStartDate()` has a mistake.
```java
	public boolean isVeseinTalUmatarStartDate() {
		if (inIsrael) {
			if (getJewishMonth() == CHESHVAN && getJewishDayOfMonth() == 7) {return true;}
		} else {
			return getTekufasTishreiElapsedDays() == 47;
		}
	}
```

**Comment by KosherJava** [comment]
@yparitcher,
Thank you very much. I hope that the latest code covers the remainder of your comments.

**Comment by yparitcher** [comment]
:+1: 
Looks Great

`isVeseinTalUmatarStartDate()` now returns the day before (the starting night) not the first day.

---

https://github.com/KosherJava/zmanim/blob/277f34f83eedd3b72bc4a9b79d1694c19269da04/src/main/java/com/kosherjava/zmanim/hebrewcalendar/JewishCalendar.java#L1178-L1186

Should probably be on line 1138 with the rest of VeseinTalUmatar.

**Comment by KosherJava** [comment]
@yparitcher , am I correct that in Israel the current code will return the actual day, while in Chutz Laaretz the code will return the day before?

**Comment by yparitcher** [comment]
> @yparitcher , am I correct that in Israel the current code will return the actual day, while in Chutz Laaretz the code will return the day before?

yes

**Comment by KosherJava** [comment]
I am on the fence if I should rename it to indicate it is the night and both should be the day earlier, or have both on the proper day and name it to indicate that. On one hand it is funny to have it the day early, on the other hand, isn't that what people putting out calendars would expect?

**Comment by yparitcher** [comment]
> I am on the fence if I should rename it to indicate it is the night and both should be the day earlier, or have both on the proper day and name it to indicate that. On one hand it is funny to have it the day early, on the other hand, isn't that what people putting out calendars would expect?

Not sure, unless you want to have both?

**Comment by KosherJava** [comment]
@yparitcher 
Freilichen Chanukah. In order to simplify returning the today and tonight versions, I think it would make sense to have a method like the following. Can you fill in the proper "else" clause below?
Thanks

```
//we can probably make this public, but it would be inconsistent with the rest of the API
private JewishDate getVeseinTalUmatarStartDate() {
	JewishDate startDate = null;
	if (inIsrael) {
		new JewishDate(getJewishYear(), CHESHVAN, 7);
	} else {
		// How do I set the date for the chutz laaretz start date? How can I leverage getTekufasTishreiElapsedDays()
	}
	if (startDate.getDayOfWeek() == 7) {
		startDate.forward(Calendar.DATE, 1);
	}
	return startDate;
}
```

**Comment by yparitcher** [comment]
You can't.
You would have to loop through days until you get to the right one.

Mashiv haruach etc. return a boolean `is..` so returning a date is a little inconsistent.

**Comment by KosherJava** [comment]
Thanks @yparitcher . I would appreciate if you could review the code (and documentation/comments) below for accuracy.

```
/**
 * Returns if it is the Jewish day (starting the evening before) to start reciting <em>Vesein Tal Umatar
 * Livracha</em> (<em>Sheailas Geshamim</em>). In Israel this is the 7th day of <em>Marcheshvan</em>. Outside
 * Israel recitation starts on the evening of December 4th (or 5th if it is the year before a civil leap year)
 * in the 21st century and shifts a day forward every century not evenly divisible by 400. This method will
 * return true if <em>vesein tal umatar</em> on the current Jewish date that starts on the previous night, so
 * Dec 5/6 will be returned by this method in the 21st century. <em>vesein tal umatar</em> is not recited on
 * <em>Shabbos</em> and the start date will be delayed a day when the start day is on a <em>Shabbos</em> (this
 * can only occur out of Israel).
 * 
 * @return true if it is the first Jewish day (starting the prior evening of reciting <em>Vesein Tal Umatar
 * Livracha</em> (<em>Sheailas Geshamim</em>).
 */
public boolean isVeseinTalUmatarStartDate() {
	if (inIsrael) {
		 // The 7th Cheshvan can't occur on Shabbos, so always return true for 7 Cheshvan
		if (getJewishMonth() == CHESHVAN && getJewishDayOfMonth() == 7) {
			return true;
		}
	} else {
		if (getDayOfWeek() == 7) { //Not recited on Friday night
			return false;
		}
		if(getDayOfWeek() == 1) { // if the start day occurs on Shabbos, return the following day
			return getTekufasTishreiElapsedDays() == 48;
		} else {
			return getTekufasTishreiElapsedDays() == 47;
		}
	}
	return false; // keep the compiler happy
}


/**
 * Returns if true if tonight is the first night to start reciting <em>Vesein Tal Umatar Livracha</em> (
 * <em>Sheailas Geshamim</em>). In Israel this is the 7th day of <em>Marcheshvan</em> (so the 6th will return
 * true). Outside Israel recitation starts on the evening of December 4th (or 5th if it is the year before a
 * civil leap year) in the 21st century and shifts a day forward every century not evenly divisible by 400.
 * <em>Vesein tal umatar</em> is not recited on <em>Shabbos</em> and the start date will be delayed a day when
 * the start day is on a <em>Shabbos</em> (this can only occur out of Israel).
 * 
 * @return true if it is the first Jewish day (starting the prior evening of reciting <em>Vesein Tal Umatar
 * Livracha</em> (<em>Sheailas Geshamim</em>).
 */
public boolean isVeseinTalUmatarStartingTonight() {
	if (inIsrael) {
		// The 7th Cheshvan can't occur on Shabbos, so always return true for 6 Cheshvan
		if (getJewishMonth() == CHESHVAN && getJewishDayOfMonth() == 6) {
				return true;
		}
	} else {
		if (getDayOfWeek() == 6) { //Not recited on Friday night
			return false;
		}
		if(getDayOfWeek() == 7) {
			// if the start day occurs on Friday night, return the following night
			return getTekufasTishreiElapsedDays() == 47;
		} else {
			return getTekufasTishreiElapsedDays() == 46;
		}
	}
	return false;
}
```

**Comment by yparitcher** [comment]
one typo, see my comments on your latest commit

**Comment by yparitcher** [comment]
@KosherJava 

Looks great. :+1:

**Comment by KosherJava** [comment]
Thanks @yparitcher . If it was still the first night of Chanukah I would have extra kavanah in Shehecheyanu 😊.

**Pull request merged**